posthog-core/src/index.ts

```diff
@@ -949,69 +964,100 @@ export abstract class PostHogCoreStateless {
+        sent_at: currentISOTime(),
+      }
 
-    const payload = JSON.stringify(data)
+      if (this.historicalMigration) {
+        data.historical_migration = true
+      }
 
-    const url =
-      this.captureMode === 'form'
-        ? `${this.host}/e/?ip=1&_=${currentTimestamp()}&v=${this.getLibraryVersion()}`
-        : `${this.host}/batch/`
+      const payload = JSON.stringify(data)
+
+      const url =
+        this.captureMode === 'form'
+          ? `${this.host}/e/?ip=1&_=${currentTimestamp()}&v=${this.getLibraryVersion()}`
+          : `${this.host}/batch/`
+
+      const fetchOptions: PostHogFetchOptions =
+        this.captureMode === 'form'
+          ? {
+              method: 'POST',
+              mode: 'no-cors',
+              credentials: 'omit',
+              headers: { ...this.getCustomHeaders(), 'Content-Type': 'application/x-www-form-urlencoded' },
+              body: `data=${encodeURIComponent(LZString.compressToBase64(payload))}&compression=lz64`,
+            }
+          : {
+              method: 'POST',
+              headers: { ...this.getCustomHeaders(), 'Content-Type': 'application/json' },
+              body: payload,
+            }
 
-    const fetchOptions: PostHogFetchOptions =
-      this.captureMode === 'form'
-        ? {
-            method: 'POST',
-            mode: 'no-cors',
-            credentials: 'omit',
-            headers: { ...this.getCustomHeaders(), 'Content-Type': 'application/x-www-form-urlencoded' },
-            body: `data=${encodeURIComponent(LZString.compressToBase64(payload))}&compression=lz64`,
+      const retryOptions: Partial<RetriableOptions> = {
+        retryCheck: (err) => {
+          // don't automatically retry on 413 errors, we want to reduce the batch size first
+          if (isPostHogFetchContentTooLargeError(err)) {
+            return false
           }
-        : {
-            method: 'POST',
-            headers: { ...this.getCustomHeaders(), 'Content-Type': 'application/json' },
-            body: payload,
-          }
-
-    try {
-      await this.fetchWithRetry(url, fetchOptions)
-    } catch (err) {
-      // depending on the error type, eg a malformed JSON or broken queue, it'll always return an error
-      // and this will be an endless loop, in this case, if the error isn't a network issue, we always remove the items from the queue
-      if (!(err instanceof PostHogFetchNetworkError)) {
-        persistQueueChange()
+          // otherwise, retry on network errors
+          return isPostHogFetchError(err)
+        },
       }
-      this._events.emit('error', err)
 
-      throw err
-    }
+      try {
+        await this.fetchWithRetry(url, fetchOptions, retryOptions)
+      } catch (err) {
+        if (isPostHogFetchContentTooLargeError(err) && batchMessages.length > 1) {
+          // if we get a 413 error, we want to reduce the batch size and try again
+          this.maxBatchSize = Math.max(1, Math.floor(batchMessages.length / 2))
```

Comment:
This changes the batch size forever, but this could have happened because of a single big message. What if we had an internal ephemeral batch size instead?

Reply:
I did wonder about that, but it would mean that we'd potentially end up needing to send each batch multiple times. It could "learn" over time (hey this sounds a bit like how TCP MTU works) but I wanted to keep it simple for now.